File: Miharu+ System - Desktop/Miharu.Data.ServiceRemote/DataBase/ProcessQuery.cs
using Miharu.Data.ServiceRemote.Model;
using System;
using System.Collections.Generic;
using System.Data;
using System.Data.SqlClient;
using System.Linq;
using System.Text.RegularExpressions;
using System.Web;

namespace Miharu.Data.ServiceRemote.DataBase
{
    public class ProcessQuery
    {
        public string getConnectionString()
        {
            return System.Configuration.ConfigurationManager.ConnectionStrings["MiharuConnectionString"].ConnectionString;
        }

        public QueryResponse execute(QueryRequest queryRequest)
        {
            if (queryRequest.parameters == null)
                queryRequest.parameters = new List<QueryParameter>();

            QueryResponse queryResponse = new QueryResponse();
            try
            {
                if (queryRequest.queryRequestType == QueryRequestType.StoredProcedure)
                {
                    queryResponse = executeStoredProcedure(queryRequest);
                }
                else
                {
                    queryResponse.dataTable = executeTableQuery(queryRequest);
                    queryResponse.rows = queryResponse.dataTable.Rows.Count;
                    queryResponse.success = true;
                }
            }
            catch (Exception ex)
            {
                queryResponse.error = ex.Message;
                queryResponse.success = false;
            }
            return queryResponse;
        }

        private QueryResponse executeStoredProcedure(QueryRequest queryRequest)
        {
            QueryResponse queryResponse = new QueryResponse();
            switch (queryRequest.queryResponseType)
            {
                case QueryResponseType.Table:
                    queryResponse.dataTable = executeDataTable(queryRequest);
                    queryResponse.rows = queryResponse.dataTable.Rows.Count;
                    queryResponse.success = true;
                    break;
                case QueryResponseType.Scalar:
                    queryResponse.scalar = executeScalar(queryRequest);
                    queryResponse.success = true;
                    break;
                case QueryResponseType.NonQuery:
                    queryResponse.rows = executeNonQuery(queryRequest);
                    queryResponse.success = true;
                    break;
                default:
                    queryResponse.error = "Tipo operacion desconocida";
                    queryResponse.success = false;
                    break;
            }
            return queryResponse;
        }

        private DataTable executeDataTable(QueryRequest queryRequest)
        {
            SqlConnection sqlConnection = null;

            try
            {
                sqlConnection = new SqlConnection(getConnectionString());
                sqlConnection.Open();
                SqlCommand sqlCommand = sqlConnection.CreateCommand();
                sqlCommand.CommandType = CommandType.StoredProcedure;
                sqlCommand.CommandText = queryRequest.name;
                SqlCommandBuilder.DeriveParameters(sqlCommand);
                foreach (QueryParameter parameter in queryRequest.parameters)
                {
                    ////object valueToAssign = Guid.TryParse(parameter.value, out Guid guidValue) ? (object)guidValue : parameter.value;
                    //object valueToAssign = (Guid.TryParse(parameter.value, out Guid guidValue) ? (object)guidValue :
                    //                                                                              (DateTime.TryParse(parameter.value, out DateTime dateValue) ? (object)dateValue : parameter.value));
                    //sqlCommand.Parameters["@" + parameter.name].Value = valueToAssign;

                    string paramName = "@" + parameter.name;
                    var sqlParam = sqlCommand.Parameters[paramName];

                    AssignParameterValueFromRawString(sqlParam, parameter.value);
                }
                SqlDataAdapter sqlDataAdapter = new SqlDataAdapter(sqlCommand);
                DataTable dataTable = new DataTable();
                sqlDataAdapter.Fill(dataTable);
                sqlDataAdapter.Dispose();
                dataTable.TableName = "Response";
                return dataTable;
            }
            finally
            {
                try
                {
                    sqlConnection.Close();
                }
                catch { }
            }
        }


        private Object executeScalar(QueryRequest queryRequest)
        {
            SqlConnection sqlConnection = null;

            try
            {
                sqlConnection = new SqlConnection(getConnectionString());
                sqlConnection.Open();
                SqlCommand sqlCommand = sqlConnection.CreateCommand();
                sqlCommand.CommandType = CommandType.StoredProcedure;
                sqlCommand.CommandText = queryRequest.name;
                SqlCommandBuilder.DeriveParameters(sqlCommand);
                foreach (QueryParameter parameter in queryRequest.parameters)
                {
                    ////object valueToAssign = Guid.TryParse(parameter.value, out Guid guidValue) ? (object)guidValue : parameter.value;
                    //object valueToAssign = (Guid.TryParse(parameter.value, out Guid guidValue) ? (object)guidValue :
                    //                                                          (DateTime.TryParse(parameter.value, out DateTime dateValue) ? (object)dateValue : parameter.value));

                    //sqlCommand.Parameters["@" + parameter.name].Value = valueToAssign;
                    //

                    string paramName = "@" + parameter.name;
                    var sqlParam = sqlCommand.Parameters[paramName];

                    AssignParameterValueFromRawString(sqlParam, parameter.value);

                    //string paramName = "@" + parameter.name;
                    //string rawValue = parameter.value;

                    //object valueToAssign = rawValue;

                    //var sqlParam = sqlCommand.Parameters[paramName];

                    //// Analizar tipo de SQL Server
                    //switch (sqlParam.SqlDbType)
                    //{
                    //    case SqlDbType.UniqueIdentifier:
                    //        if (Guid.TryParse(rawValue, out Guid guidVal))
                    //            valueToAssign = guidVal;
                    //        else
                    //            valueToAssign = DBNull.Value;
                    //        break;

                    //    case SqlDbType.DateTime:
                    //    case SqlDbType.Date:
                    //        if (DateTime.TryParse(rawValue, out DateTime dateVal))
                    //            valueToAssign = dateVal;
                    //        else
                    //            valueToAssign = DBNull.Value;
                    //        break;

                    //    case SqlDbType.VarBinary:
                    //        try
                    //        {
                    //            // Convertir Base64 a byte[]
                    //            valueToAssign = Convert.FromBase64String(rawValue);
                    //        }
                    //        catch
                    //        {
                    //            valueToAssign = DBNull.Value;
                    //        }
                    //        break;

                    //    default:
                    //        valueToAssign = rawValue;
                    //        break;
                    //}

                    //sqlParam.Value = valueToAssign ?? DBNull.Value;
                }

                return sqlCommand.ExecuteScalar();
            }
            finally
            {
                try
                {
                    sqlConnection.Close();
                }
                catch { }
            }
        }

        private int executeNonQuery(QueryRequest queryRequest)
        {
            SqlConnection sqlConnection = null;

            try
            {
                sqlConnection = new SqlConnection(getConnectionString());
                sqlConnection.Open();
                SqlCommand sqlCommand = sqlConnection.CreateCommand();
                sqlCommand.CommandType = CommandType.StoredProcedure;
                sqlCommand.CommandText = queryRequest.name;
                SqlCommandBuilder.DeriveParameters(sqlCommand);
                foreach (QueryParameter parameter in queryRequest.parameters)
                {
                    ////object valueToAssign = Guid.TryParse(parameter.value, out Guid guidValue) ? (object)guidValue : parameter.value;
                    //object valueToAssign = (Guid.TryParse(parameter.value, out Guid guidValue) ? (object)guidValue :
                    //                                                          (DateTime.TryParse(parameter.value, out DateTime dateValue) ? (object)dateValue : parameter.value));
                    //sqlCommand.Parameters["@" + parameter.name].Value = valueToAssign;

                    string paramName = "@" + parameter.name;
                    var sqlParam = sqlCommand.Parameters[paramName];

                    AssignParameterValueFromRawString(sqlParam, parameter.value);
                }
                return sqlCommand.ExecuteNonQuery();
            }
            finally
            {
                try
                {
                    sqlConnection.Close();
                }
                catch { }
            }
        }

        private DataTable executeTableQuery(QueryRequest queryRequest)
        {
            SqlConnection sqlConnection = null;

            string sqlstring = "select * from " + queryRequest.name;

            if (queryRequest.parameters.Count > 0)
            {
                sqlstring += " where ";
                int i = 0;
                foreach (QueryParameter parameter in queryRequest.parameters)
                {
                    if (i > 0)
                    {
                        sqlstring += " and ";
                    }
                    sqlstring += "[" + parameter.name + "]=@" + parameter.name;
                    i++;
                }
            }

            try
            {
                sqlConnection = new SqlConnection(getConnectionString());
                sqlConnection.Open();
                SqlCommand sqlCommand = sqlConnection.CreateCommand();
                sqlCommand.CommandType = CommandType.Text;
                sqlCommand.CommandText = sqlstring;
                foreach (QueryParameter parameter in queryRequest.parameters)
                {
                    sqlCommand.Parameters.Add(new SqlParameter("@" + parameter.name, parameter.value));
                }
                SqlDataAdapter sqlDataAdapter = new SqlDataAdapter(sqlCommand);
                DataTable dataTable = new DataTable();
                sqlDataAdapter.Fill(dataTable);
                sqlDataAdapter.Dispose();
                dataTable.TableName = "Response";
                return dataTable;
            }
            finally
            {
                try
                {
                    sqlConnection.Close();
                }
                catch { }
            }
        }

        private void AssignParameterValueFromRawString(SqlParameter sqlParam, string rawValue)
        {
            if (string.IsNullOrWhiteSpace(rawValue))
            {
                sqlParam.Value = DBNull.Value;
                return;
            }

            try
            {
                switch (sqlParam.SqlDbType)
                {
                    case SqlDbType.UniqueIdentifier:
                        sqlParam.Value = Guid.TryParse(rawValue, out Guid guidVal) ? (object)guidVal : DBNull.Value;
                        break;

                    case SqlDbType.DateTime:
                    case SqlDbType.Date:
                        sqlParam.Value = DateTime.TryParse(rawValue, out DateTime dateVal) ? (object)dateVal : DBNull.Value;
                        break;

                    case SqlDbType.VarBinary:
                        sqlParam.Value = Convert.FromBase64String(rawValue);
                        break;

                    default:
                        sqlParam.Value = rawValue;
                        break;
                }
            }
            catch
            {
                sqlParam.Value = DBNull.Value;
            }
        }
    }
}

File: Miharu+ System - Imaging/Miharu.Imaging.Explorer.WebService/Utils/ProcessQuery.vb
Imports System
Imports System.Collections.Generic
Imports System.Data
Imports System.Data.SqlClient
Imports System.Linq
Imports System.Web

Public Class ProcessQuery
    Public Function getConnectionString() As String
        Return System.Configuration.ConfigurationManager.ConnectionStrings("SQLCnnMiharu.CoreDMZ").ConnectionString
    End Function

    Public Function getConnectionStringImaging() As String
        Return System.Configuration.ConfigurationManager.ConnectionStrings("SQLCnnMiharu.ImagingDMZ").ConnectionString
    End Function

    Public Function execute(ByVal queryRequest As QueryRequest) As QueryResponse
        If queryRequest.parameters Is Nothing Then queryRequest.parameters = New List(Of QueryParameter)()
        Dim queryResponse As QueryResponse = New QueryResponse()

        Try

            If queryRequest.queryRequestType = QueryRequestType.StoredProcedure Then
                queryResponse = executeStoredProcedure(queryRequest)
            Else
                queryResponse.dataTable = executeTableQuery(queryRequest)
                queryResponse.rows = queryResponse.dataTable.Rows.Count
                queryResponse.success = True
            End If

        Catch ex As Exception
            queryResponse.[error] = ex.Message
            queryResponse.success = False
        End Try

        Return queryResponse
    End Function

    Private Function executeStoredProcedure(ByVal queryRequest As QueryRequest) As QueryResponse
        Dim queryResponse As QueryResponse = New QueryResponse()

        Select Case queryRequest.queryResponseType
            Case QueryResponseType.Table
                queryResponse.dataTable = executeDataTable(queryRequest)
                queryResponse.rows = queryResponse.dataTable.Rows.Count
                queryResponse.success = True
            Case QueryResponseType.Scalar
                queryResponse.scalar = executeScalar(queryRequest)
                queryResponse.success = True
            Case QueryResponseType.NonQuery
                queryResponse.rows = executeNonQuery(queryRequest)
                queryResponse.success = True
            Case Else
                queryResponse.[error] = "Tipo operacion desconocida"
                queryResponse.success = False
        End Select

        Return queryResponse
    End Function

    Private Function executeDataTable(ByVal queryRequest As QueryRequest) As DataTable
        Dim sqlConnection As SqlConnection = Nothing

        Try
            sqlConnection = New SqlConnection(getConnectionString())
            sqlConnection.Open()
            Dim sqlCommand As SqlCommand = sqlConnection.CreateCommand()
            sqlCommand.CommandType = CommandType.StoredProcedure
            sqlCommand.CommandText = queryRequest.name
            SqlCommandBuilder.DeriveParameters(sqlCommand)

            For Each parameter As QueryParameter In queryRequest.parameters
                sqlCommand.Parameters("@" & parameter.name).Value = parameter.value
            Next

            Dim sqlDataAdapter As SqlDataAdapter = New SqlDataAdapter(sqlCommand)
            Dim dataTable As DataTable = New DataTable()
            sqlDataAdapter.Fill(dataTable)
            sqlDataAdapter.Dispose()
            dataTable.TableName = "Response"
            Return dataTable
        Finally

            Try
                sqlConnection.Close()
            Catch
            End Try
        End Try
    End Function

    Private Function executeScalar(ByVal queryRequest As QueryRequest) As Object
        Dim sqlConnection As SqlConnection = Nothing

        Try
            sqlConnection = New SqlConnection(getConnectionString())
            sqlConnection.Open()
            Dim sqlCommand As SqlCommand = sqlConnection.CreateCommand()
            sqlCommand.CommandType = CommandType.StoredProcedure
            sqlCommand.CommandText = queryRequest.name
            SqlCommandBuilder.DeriveParameters(sqlCommand)

            For Each parameter As QueryParameter In queryRequest.parameters
                sqlCommand.Parameters("@" & parameter.name).Value = parameter.value
            Next

            Return sqlCommand.ExecuteScalar()
        Finally

            Try
                sqlConnection.Close()
            Catch
            End Try
        End Try
    End Function

    Private Function executeNonQuery(ByVal queryRequest As QueryRequest) As Integer
        Dim sqlConnection As SqlConnection = Nothing

        Try
            sqlConnection = New SqlConnection(getConnectionString())
            sqlConnection.Open()
            Dim sqlCommand As SqlCommand = sqlConnection.CreateCommand()
            sqlCommand.CommandType = CommandType.StoredProcedure
            sqlCommand.CommandText = queryRequest.name
            SqlCommandBuilder.DeriveParameters(sqlCommand)

            For Each parameter As QueryParameter In queryRequest.parameters
                sqlCommand.Parameters("@" & parameter.name).Value = parameter.value
            Next

            Return sqlCommand.ExecuteNonQuery()
        Finally

            Try
                sqlConnection.Close()
            Catch
            End Try
        End Try
    End Function

    Private Function executeTableQuery(ByVal queryRequest As QueryRequest) As DataTable
        Dim sqlConnection As SqlConnection = Nothing
        Dim sqlstring As String = "select * from " & queryRequest.name

        If queryRequest.parameters.Count > 0 Then
            sqlstring += " where "
            Dim i As Integer = 0

            For Each parameter As QueryParameter In queryRequest.parameters

                If i > 0 Then
                    sqlstring += " and "
                End If

                sqlstring += "[" & parameter.name & "]=@" + parameter.name
                i += 1
            Next
        End If

        Try
            sqlConnection = New SqlConnection(getConnectionString())
            sqlConnection.Open()
            Dim sqlCommand As SqlCommand = sqlConnection.CreateCommand()
            sqlCommand.CommandType = CommandType.Text
            sqlCommand.CommandText = sqlstring

            For Each parameter As QueryParameter In queryRequest.parameters
                sqlCommand.Parameters.Add(New SqlParameter("@" & parameter.name, parameter.value))
            Next

            Dim sqlDataAdapter As SqlDataAdapter = New SqlDataAdapter(sqlCommand)
            Dim dataTable As DataTable = New DataTable()
            sqlDataAdapter.Fill(dataTable)
            sqlDataAdapter.Dispose()
            dataTable.TableName = "Response"
            Return dataTable
        Finally

            Try
                sqlConnection.Close()
            Catch
            End Try
        End Try
    End Function
End Class
